Use newer version of Terraform AWS provider

jsf9k opened the pull request:
## 🗣 Description ##

This pull request:
- Upgrades the `terraform` code to use a newer version of the Terraform AWS provider.
- Commits the `.terraform.lock.hcl` file for more reproducible builds.
- Upgrades the required version of the `terraform` executable since we use the `nullable` attribute for Terraform variables.
- Adds a `README.md` file containing the `terraform-docs` output for the `terraform` directory.

## 💭 Motivation and context ##

We should use a modern Terraform AWS provider wherever we can; otherwise, we will get left behind.

Adding the `.terraform.lock.hcl` file is a recommended practice, and it will lead to more reproducible builds.  See [the Terraform documentation](https://developer.hashicorp.com/terraform/language/files/dependency-lock) for more details.

## 🧪 Testing ##

All automated tests pass.  I also ran `terraform apply` in the `terraform` directory for all three COOL deployments and verified that nothing changed other than the addition of the `region` attribute to the AMI objects in the outputs.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

jsf9k committed:
Require at least version 1.1 of Terraform

Version 1.1 of Terraform is the first version to support the nullable
key in variable definitions, and we are making use of nullable in this
Terraform code.

jsf9k committed:
Use the latest AWS provider for Terraform

I verified that this version works for this repository by re-running
terraform apply for our dev-a, staging-a, and production deployments.

jsf9k commented:
@mcdonnnj - Is a version bump warranted here?

mcdonnnj reviewed (approved):
LGTM ✔ Since there are no changes to the Ansible role itself I would not say a version bump is warranted here.

jsf9k committed:
Remove .terraform.lock.hcl from .gitignore

jsf9k committed:
Add .terraform.lock.hcl file

Using this file results in more reproducible builds and protects us
against the insertion of a malicious provider.  See
https://developer.hashicorp.com/terraform/language/files/dependency-lock
for more details.

jsf9k committed:
Add README.md containing terraform-docs output for terraform directory

mcdonnnj commented on terraform/README.md:
```diff
@@ -0,0 +1,55 @@
+that is used to create a test user.  This test user is in turn used by
+our CI/CD pipeline to test the Ansible role.
+
+See [the overall project documentation](../README.md) for a detailed
```
I prefer we leave articles outside of links.

```suggestion
See the [overall project documentation](../README.md) for a detailed
```

mcdonnnj reviewed (approved):
One item of feedback for your consideration.

jsf9k committed:
Remove article from link text

@mcdonnnj finds this distasteful.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
